### PR TITLE
Check if the scheduler already has `reduce_on_plateau`

### LINKED ISSUE #13569: Support for returning LRSchedulerConfig on LightningModule.configure_optimizers
## 🚀 Feature

Currently, LRSchedulerConfig type is not supported by [_configure_schedulers_automatic_opt](https://github.com/Lightning-AI/lightning/blob/61c28cb428a13c2aea6d7f3f55e0f00431a4ea4e/src/pytorch_lightning/core/optimizer.py#L263). 

### Motivation

I created a custom `ReduceLROnPlateau` class that adds warmup epochs before `ReduceLROnPlateau`. However, the `ReduceLROnPlateau` scheduler is only supported for the `torch.optim.lr_scheduler.ReduceLROnPlateau` class. This is [implemented](https://github.com/Lightning-AI/lightning/blob/61c28cb428a13c2aea6d7f3f55e0f00431a4ea4e/src/pytorch_lightning/core/optimizer.py#L288) by setting `reduce_on_plateau` property based on `isinstance` check based on the type of the scheduler. The check causes a problem using a custom scheduling class with similar properties as 'ReduceLROnPlateau'. There is a need for an alternative mechanism to notify the `_configure_schedulers_automatic_opt` that the scheduler is a derivative of ReduceLROnPlateau (not necessarily a subclass).

### Pitch

If desired, users can gain more control by bypassing this mechanism by directly returning `LRSchedulerConfig` from `configure_optimizers` rather than a dictionary. This can be implemented by simple two-line addition to [here](https://github.com/Lightning-AI/lightning/blob/61c28cb428a13c2aea6d7f3f55e0f00431a4ea4e/src/pytorch_lightning/core/optimizer.py#L311):

```python
...
elif isinstance(scheduler, LRSchedulerConfig):
    config = scheduler
...
```

This is a very straightforward solution to this problem. I'm not aware of any side effects. It is fair to assume it is an advanced use case and does not break any compatibility with the existing code base. 

cc @borda @rohitgr7 @akihironitta

### PR DESCRIPTION
**Pull request by MrShevan**
## What does this PR do?

Fixes #13569

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: src/pytorch_lightning/core/optimizer.py [hunk 1]
```diff
@@ -284,7 +284,15 @@ def _configure_schedulers_automatic_opt(schedulers: list, monitor: Optional[str]
...
-            scheduler["reduce_on_plateau"] = isinstance(scheduler["scheduler"], optim.lr_scheduler.ReduceLROnPlateau)
+            if "reduce_on_plateau" not in scheduler:
+                scheduler["reduce_on_plateau"] = isinstance(
+                    scheduler["scheduler"], optim.lr_scheduler.ReduceLROnPlateau
+                )
+            elif not isinstance(scheduler["reduce_on_plateau"], bool):
+                raise MisconfigurationException(
+                    'The "reduce_on_plateau" key in lr scheduler dict must be a bool'
+                    f' but is "{type(scheduler["reduce_on_plateau"])}"'
+                )
```
Comment (rohitgr7) [inline src/pytorch_lightning/core/optimizer.py:295]: I think this is not required since it's typed.

File: src/pytorch_lightning/core/optimizer.py [hunk 2]
```diff
@@ -284,7 +284,15 @@ def _configure_schedulers_automatic_opt(schedulers: list, monitor: Optional[str]
                     'The "interval" key in lr scheduler dict must be "step" or "epoch"'
                     f' but is "{scheduler["interval"]}"'
                 )
-            scheduler["reduce_on_plateau"] = isinstance(scheduler["scheduler"], optim.lr_scheduler.ReduceLROnPlateau)
+            if "reduce_on_plateau" not in scheduler:
+                scheduler["reduce_on_plateau"] = isinstance(
+                    scheduler["scheduler"], optim.lr_scheduler.ReduceLROnPlateau
+                )
```
Comment (rohitgr7) [inline src/pytorch_lightning/core/optimizer.py:290]:
```suggestion
            scheduler["reduce_on_plateau"] = scheduler.get(
                    "reduce_on_plateau",
                    isinstance(scheduler["scheduler"], optim.lr_scheduler.ReduceLROnPlateau)
            )
```

File: src/pytorch_lightning/core/optimizer.py [hunk 3]
```diff
@@ -284,7 +284,15 @@ def _configure_schedulers_automatic_opt(schedulers: list, monitor: Optional[str]
                     'The "interval" key in lr scheduler dict must be "step" or "epoch"'
                     f' but is "{scheduler["interval"]}"'
                 )
-            scheduler["reduce_on_plateau"] = isinstance(scheduler["scheduler"], optim.lr_scheduler.ReduceLROnPlateau)
+            if "reduce_on_plateau" not in scheduler:
```
Comment (rohitgr7) [inline src/pytorch_lightning/core/optimizer.py:287]: mind add a similar unittest covering this case?

**Review by rohitgr7 — COMMENTED** [review]
Mind add a changelog entry?

File: src/pytorch_lightning/CHANGELOG.md [hunk 4]
```diff
@@ -376,6 +376,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Fixed main progress bar counter when `val_check_interval=int` and `check_val_every_n_epoch=None` ([#12832](https://github.com/Lightning-AI/lightning/pull/12832)
 
 
+- Improved support for custom scheduler `ReduceLROnPlateau` class ([#13569](https://github.com/Lightning-AI/lightning/issues/13569))
```
Comment (rohitgr7) [inline src/pytorch_lightning/CHANGELOG.md:379]:
```suggestion
- Improved support for custom `ReduceLROnPlateau` scheduler if `reduce_on_plateau` is set by the user in scheduler config ([#13838](https://github.com/Lightning-AI/lightning/pull/13838))
```

File: tests/tests_pytorch/trainer/optimization/test_optimizers.py [hunk 5]
```diff
@@ -115,6 +115,17 @@ def test_onecyclelr_with_epoch_interval_warns():
         _configure_schedulers_automatic_opt([lr_scheduler], None)
 
 
+def test_scheduler_initialized_with_custom_reduceonplateau():
+    """Test for initialize custom scheduler with `reduce_on_plateau`argument."""
```
Comment (rohitgr7) [inline tests/tests_pytorch/trainer/optimization/test_optimizers.py:119]:
```suggestion
    """Test for initialize custom scheduler with `reduce_on_plateau` argument."""
```

File: tests/tests_pytorch/trainer/optimization/test_optimizers.py [hunk 6]
```diff
@@ -115,6 +115,17 @@ def test_onecyclelr_with_epoch_interval_warns():
...
+
+    class CustomReduceLROnPlateau:
+        pass
+
+    lr_scheduler = {"reduce_on_plateau": True, "scheduler": CustomReduceLROnPlateau(), "monitor": "my_loss"}
+    config = _configure_schedulers_automatic_opt([lr_scheduler], None)
+    assert isinstance(config[0].scheduler, CustomReduceLROnPlateau)
```
Comment (rohitgr7) [inline tests/tests_pytorch/trainer/optimization/test_optimizers.py:126]:
```suggestion
    assert isinstance(config[0].scheduler, CustomReduceLROnPlateau)
    assert config[0].reduce_on_plateau is True
```

**Review by rohitgr7 — APPROVED** [review]
Neat

File: tests/tests_pytorch/trainer/optimization/test_optimizers.py [hunk 7]
```diff
@@ -115,6 +115,18 @@ def test_onecyclelr_with_epoch_interval_warns():
...
+def test_scheduler_initialized_with_custom_reduceonplateau():
+    """Test for initialize custom scheduler with `reduce_on_plateau` argument."""
+
+    class CustomReduceLROnPlateau:
+        pass
+
+    lr_scheduler = {"reduce_on_plateau": True, "scheduler": CustomReduceLROnPlateau(), "monitor": "my_loss"}
+    config = _configure_schedulers_automatic_opt([lr_scheduler], None)
+    assert isinstance(config[0].scheduler, CustomReduceLROnPlateau)
+    assert config[0].reduce_on_plateau is True
```
Comment (carmocca) [inline tests/tests_pytorch/trainer/optimization/test_optimizers.py:127]:
```suggestion
    assert config[0].reduce_on_plateau
```

**Review by carmocca — APPROVED** [review]

**Review by Borda — APPROVED** [review]
